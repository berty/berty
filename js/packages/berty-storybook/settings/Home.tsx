import React from 'react'
import { View, ScrollView, StyleSheet, ActivityIndicator, SafeAreaView } from 'react-native'
import { Text } from 'react-native-ui-kitten'
import { useStyles } from '@berty-tech/styles'
import { ButtonSetting, ButtonSettingRow } from '../shared-components/SettingsButtons'
import { ProceduralCircleAvatar } from '../shared-components/ProceduralCircleAvatar'
import HeaderSettings from '../shared-components/Header'
import { BertyChatChatService as Store } from '@berty-tech/berty-store'
import { ScreenProps, useNavigation } from '@berty-tech/berty-navigation'
import { berty } from '@berty-tech/api'
import { Chat } from '@berty-tech/hooks'

//
// Home Vue
//

// Type

// Style
const useStylesHome = () => {
	const [{ width, height, margin, padding, text }] = useStyles()
	return {
		homeAvatarBox: [width(160), height(180)],
		firstHeaderButton: [margin.right.scale(20), height(90)],
		secondHeaderButton: [margin.right.scale(20), height(90)],
		thirdHeaderButton: height(90),
		headerNameText: text.size.scale(13),
		scrollViewPadding: padding.bottom.scale(116),
	}
}
const _homeStyles = StyleSheet.create({
	homeAvatar: { bottom: 37.5 },
})

const HomeHeaderGroupButton: React.FC<berty.chatmodel.Account> = () => {
	const _styles = useStylesHome()
	const [{ padding, color }] = useStyles()
	const { navigate } = useNavigation()
	return (
		<View style={[padding.horizontal.medium]}>
			<ButtonSettingRow
				state={[
					{
						name: 'Updates',
						icon: 'arrow-upward-outline',
						color: color.blue,
						style: _styles.firstHeaderButton,
						onPress: navigate.settings.appUpdates,
					},
					{
						name: 'Help',
						icon: 'question-mark-circle-outline',
						color: color.red,
						style: _styles.secondHeaderButton,
						onPress: navigate.settings.help,
					},
					{
						name: 'Settings',
						icon: 'settings-2-outline',
						color: color.blue,
						style: _styles.thirdHeaderButton,
						onPress: navigate.settings.mode,
					},
				]}
			/>
		</View>
	)
}
const HomeHeaderAvatar: React.FC<berty.chatmodel.Account> = ({ contact }) => {
	const _styles = useStylesHome()
	const [{ row, margin, background, border, text, padding }] = useStyles()
	const client = Chat.useClient()
	const account = Chat.useAccount()
	return (
		<View style={[row.center, margin.top.scale(50)]}>
			<View style={[_styles.homeAvatarBox, background.white, border.radius.medium]}>
				<View style={[_homeStyles.homeAvatar]}>
					<ProceduralCircleAvatar
						style={row.center}
						seed={client?.accountPk}
						size={75}
						diffSize={25}
					/>
					<View style={[row.center]}>
						<Text style={[text.family, padding.top.small, _styles.headerNameText]}>
							{account?.name || ''}
						</Text>
					</View>
				</View>
			</View>
		</View>
	)
}

const HomeHeader: React.FC = () => {
	const [{ margin }] = useStyles()
	return (
		<SafeAreaView style={[margin.bottom.medium]}>
			<HomeHeaderAvatar />
		</SafeAreaView>
	)
}

const HomeBodySettings: React.FC<{}> = () => {
	const [{ flex, color, padding, margin }] = useStyles()
	const { navigate } = useNavigation()
	return (
		<View style={[flex.tiny, padding.horizontal.medium, margin.top.medium]}>
			<ButtonSetting
				name='Notifications'
				icon='bell-outline'
				state={{ value: 'Current', color: color.white, bgColor: color.blue }}
				onPress={navigate.settings.notifications}
			/>
			<ButtonSetting
				name='Bluetooth'
				icon='bluetooth-outline'
				onPress={navigate.settings.bluetooth}
			/>
			<ButtonSetting name='Dark mode' icon='moon-outline' toggled disabled />
			<ButtonSetting
				name='About Berty'
				icon='info-outline'
				onPress={navigate.settings.aboutBerty}
			/>
			<ButtonSetting
				name='DevTools'
				icon='options-2-outline'
				onPress={navigate.settings.devTools}
			/>
		</View>
	)
}

export const useAccount = () => {
	const [accountGet, accountGetError] = Store.useAccountGet({ id: 0 })
	const [contactGet, contactGetError] = Store.useContactGet({
		id: accountGet?.account?.contactId || -1,
	})
	if (accountGetError || accountGet == null || accountGet?.account == null) {
		accountGetError && console.error(accountGetError)
		return null
	}
	if (contactGetError || contactGet?.contact == null) {
		contactGetError && console.error(contactGetError)
		return null
	}
	return {
		...accountGet.account,
		contact: {
			...contactGet.contact,
		},
	}
}

export const Home: React.FC<ScreenProps.Settings.Home> = () => {
	const { navigate } = useNavigation()
	const account = { ...Chat.useAccount(), ...useAccount() }
	const _styles = useStylesHome()
	const [{ flex, background, row }] = useStyles()
	return (
		<View style={[flex.tiny, background.white]}>
			{account == null ? (
				<ActivityIndicator size='large' style={[row.center]} />
			) : (
				<ScrollView contentContainerStyle={[_styles.scrollViewPadding]}>
					<HeaderSettings>
						<View>
							<HomeHeader {...account} />
							<HomeHeaderGroupButton {...account} />
						</View>
					</HeaderSettings>
					<HomeBodySettings />
				</ScrollView>
			)}
		</View>
	)
}
