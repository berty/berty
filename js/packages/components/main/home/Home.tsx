import React, { useMemo, useRef, useState } from 'react'
import { useNavigation as useNativeNavigation } from '@react-navigation/native'
import { Translation } from 'react-i18next'
import { ScrollView, Text as TextNative, View } from 'react-native'
import { EdgeInsets, SafeAreaConsumer } from 'react-native-safe-area-context'
import pickBy from 'lodash/pickBy'

import { ScreenProps } from '@berty-tech/navigation'
import {
	useConversationsCount,
	useIncomingContactRequests,
	useMsgrContext,
	useNotificationsInhibitor,
	useSortedConversationList,
} from '@berty-tech/store/hooks'
import beapi from '@berty-tech/api'
import { useStyles } from '@berty-tech/styles'
import { AddBot } from '@berty-tech/components/modals'

import { useLayout } from '../../hooks'
import { SwipeNavRecognizer } from '../../common/SwipeNavRecognizer'
import BlurView from '../../common/BlurView'
import EmptyChat from '../empty_chat.svg'
import { HomeModal } from './HomeModal'
import { Footer } from './Footer'
import { IncomingRequests } from './Requests'
import { Conversations } from './Conversations'
import { SearchComponent } from './Search'
import { HomeHeader } from './Header'
import { MultiAccount } from './MultiAccount'

const T = beapi.messenger.StreamEvent.Notified.Type

export const Home: React.FC<ScreenProps.Main.Home> = () => {
	useNotificationsInhibitor((_ctx, notif) =>
		[T.TypeMessageReceived, T.TypeContactRequestReceived, T.TypeContactRequestSent].includes(
			notif.type as any,
		)
			? 'sound-only'
			: false,
	)
	// TODO: do something to animate the requests
	const requests: any[] = useIncomingContactRequests()
	const conversations: any[] = useSortedConversationList()
	const isConversation: number = useConversationsCount()
	const [layoutRequests, onLayoutRequests] = useLayout()
	const [, onLayoutHeader] = useLayout()
	const [, onLayoutConvs] = useLayout()
	const [isOnTop, setIsOnTop] = useState<boolean>(false)
	const [searchText, setSearchText] = useState<string>('')
	const [refresh, setRefresh] = useState<boolean>(false)
	const [isModalVisible, setModalVisibility] = useState<boolean>(false)
	const [isAddBot, setIsAddBot] = useState({
		link: '',
		displayName: '',
		isVisible: false,
	})
	const [isLongPress, setIsLongPress] = useState<boolean>(false)

	const { navigate } = useNativeNavigation()

	const [
		{ text, opacity, flex, margin, background },
		{ windowHeight, scaleSize, scaleHeight },
	] = useStyles()
	const scrollRef = useRef<ScrollView>(null)

	const searching = !!searchText
	const lowSearchText = searchText.toLowerCase()
	const searchCheck = React.useCallback(
		(searchIn?: string | null | false | 0) =>
			(searchIn || '').toLowerCase().includes(lowSearchText),
		[lowSearchText],
	)

	const ctx = useMsgrContext()
	const suggestions = Object.values(ctx.persistentOptions?.suggestions).filter(
		(i: any) => i.state === 'unread',
	)
	const hasSuggestion: number = suggestions.length

	const searchConversations = React.useMemo(
		() =>
			searching
				? pickBy(
						ctx.conversations,
						(val: any) =>
							val.type === beapi.messenger.Conversation.Type.MultiMemberType &&
							searchCheck(val.displayName),
				  )
				: {},
		[ctx.conversations, searchCheck, searching],
	)

	const searchContacts = React.useMemo(
		() => (searching ? pickBy(ctx.contacts, (val: any) => searchCheck(val.displayName)) : {}),
		[ctx.contacts, searchCheck, searching],
	)

	const searchInteractions = React.useMemo(() => {
		if (!searching) {
			return {}
		}
		const allInteractions: any = Object.values(ctx.interactions).reduce(
			(r: any, intes: any) => ({ ...r, ...intes }),
			{},
		)
		return pickBy(
			allInteractions,
			(inte) =>
				inte.type === beapi.messenger.AppMessage.Type.TypeUserMessage &&
				searchCheck(inte.payload?.body),
		)
	}, [ctx.interactions, searchCheck, searching])

	const hasResults = [searchConversations, searchContacts, searchInteractions].some(
		(c) => Object.keys(c).length > 0,
	)
	const styleBackground = useMemo(
		() => (requests.length > 0 && !searchText?.length ? background.blue : background.white),
		[background.blue, background.white, requests.length, searchText],
	)

	return (
		<>
			<Translation>
				{(t: any): React.ReactNode => (
					<View style={[flex.tiny, styleBackground]}>
						<SwipeNavRecognizer onSwipeLeft={() => !isModalVisible && navigate('Settings.Home')}>
							<SafeAreaConsumer>
								{(insets: EdgeInsets | null) => (
									<>
										<ScrollView
											ref={scrollRef}
											stickyHeaderIndices={!searchText?.length && !hasResults ? [1] : [0]}
											showsVerticalScrollIndicator={false}
											scrollEventThrottle={16}
											keyboardShouldPersistTaps={'handled'}
											onScrollEndDrag={(e) => {
												if (e.nativeEvent.contentOffset.y < 0) {
													setRefresh(true)
												}
											}}
											onScroll={(e) => {
												if (e.nativeEvent.contentOffset) {
													if (e.nativeEvent.contentOffset.y >= layoutRequests.height) {
														setIsOnTop(true)
													} else {
														setIsOnTop(false)
													}
												}
											}}
										>
											{!searchText?.length && (
												<IncomingRequests items={requests} onLayout={onLayoutRequests} />
											)}
											<HomeHeader
												isOnTop={isOnTop}
												hasRequests={requests.length > 0}
												scrollRef={scrollRef}
												onLayout={onLayoutHeader}
												value={searchText}
												onChange={setSearchText}
												refresh={refresh}
												setRefresh={setRefresh}
												onLongPress={setIsLongPress}
												isMultiAccount={isLongPress}
											/>
											{searchText?.length ? (
												<SearchComponent
													insets={insets}
													conversations={searchConversations}
													contacts={searchContacts}
													interactions={searchInteractions}
													value={searchText}
													hasResults={hasResults}
												/>
											) : (
												<>
													<Conversations
														items={conversations}
														suggestions={suggestions}
														onLayout={onLayoutConvs}
														addBot={setIsAddBot}
													/>
													{!isConversation && !hasSuggestion && (
														<View style={[background.white]}>
															<View
																style={[
																	flex.justify.center,
																	flex.align.center,
																	margin.top.scale(60),
																]}
															>
																<View>
																	<EmptyChat width={350 * scaleSize} height={350 * scaleHeight} />
																	<TextNative
																		style={[
																			text.align.center,
																			text.color.grey,
																			text.bold.small,
																			opacity(0.3),
																			margin.top.big,
																		]}
																	>
																		{t('main.home.no-contacts')}
																	</TextNative>
																</View>
															</View>
														</View>
													)}
													{requests.length > 0 && (
														<View
															style={[
																{
																	backgroundColor: 'white',
																	position: 'absolute',
																	bottom: windowHeight * -1,
																	height: windowHeight,
																	width: '100%',
																},
															]}
														/>
													)}
												</>
											)}
										</ScrollView>
										{isModalVisible && (
											<>
												<BlurView
													blurType='light'
													style={{
														position: 'absolute',
														left: 0,
														right: 0,
														top: 0,
														bottom: 0,
														height: windowHeight,
													}}
												/>
												<HomeModal closeModal={() => setModalVisibility(false)} />
											</>
										)}

										{!searchText?.length && (
											<Footer
												openModal={() => setModalVisibility(true)}
												isModalVisible={isModalVisible}
											/>
										)}
										{isAddBot.isVisible && (
											<AddBot
												link={isAddBot.link}
												displayName={isAddBot.displayName}
												closeModal={() => setIsAddBot({ ...isAddBot, isVisible: false })}
											/>
										)}
										{isLongPress && (
											<MultiAccount
												onPress={() => {
													setIsLongPress(false)
												}}
											/>
										)}
									</>
								)}
							</SafeAreaConsumer>
						</SwipeNavRecognizer>
					</View>
				)}
			</Translation>
		</>
	)
}

export default Home
