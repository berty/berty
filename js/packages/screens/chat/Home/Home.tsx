import pickBy from 'lodash/pickBy'
import React, { useEffect, useMemo, useRef, useState } from 'react'
import { useTranslation } from 'react-i18next'
import { ScrollView, View, StatusBar, SafeAreaView } from 'react-native'
import { useSelector } from 'react-redux'

import beapi from '@berty/api'
import EmptyChat from '@berty/assets/logo/empty_chat.svg'
import { PrimaryFloatingButton } from '@berty/components'
import { useLayout } from '@berty/components/hooks'
import { ButtonSettingV2 } from '@berty/components/shared-components'
import { UnifiedText } from '@berty/components/shared-components/UnifiedText'
import { useAppDimensions } from '@berty/contexts/app-dimensions.context'
import { useStyles } from '@berty/contexts/styles'
import {
	useContactsDict,
	useConversationsDict,
	useIncomingContactRequests,
	useAllConversations,
} from '@berty/hooks'
import { ScreenFC } from '@berty/navigation'
import { selectPersistentOptions } from '@berty/redux/reducers/persistentOptions.reducer'
import { useMessengerClient, useNotificationsInhibitor, useThemeColor } from '@berty/store'

import { AddBot } from './components/AddBot'
import { Conversations } from './components/Conversations'
import { HomeHeader } from './components/Header'
import { MultiAccount } from './components/MultiAccount'
import { IncomingRequests } from './components/Requests'
import { SearchComponent } from './components/Search'

const T = beapi.messenger.StreamEvent.Notified.Type

export const Home: ScreenFC<'Chat.Home'> = ({ navigation: { navigate } }) => {
	useNotificationsInhibitor(notif =>
		[
			T.TypeMessageReceived,
			T.TypeContactRequestReceived,
			T.TypeContactRequestSent,
			T.TypeGroupInvitation,
		].includes(notif.type || T.Unknown)
			? 'sound-only'
			: false,
	)
	// TODO: do something to animate the requests
	const requests = useIncomingContactRequests()
	const conversations = useAllConversations()
	const hasConversations = conversations.length > 0
	const [layoutRequests, onLayoutRequests] = useLayout()
	const [isOnTop, setIsOnTop] = useState<boolean>(false)
	const [searchText, setSearchText] = useState<string>('')
	const [refresh, setRefresh] = useState<boolean>(false)
	const [isAddBot, setIsAddBot] = useState({
		link: '',
		displayName: '',
		isVisible: false,
	})

	const [isLongPress, setIsLongPress] = useState<boolean>(false)

	const messengerClient = useMessengerClient()

	const { text, opacity, flex, margin, border } = useStyles()
	const { scaleSize, scaleHeight } = useAppDimensions()
	const colors = useThemeColor()
	const { t }: any = useTranslation()

	const scrollRef = useRef<ScrollView>(null)
	const searching = !!searchText
	const lowSearchText = searchText.toLowerCase()
	const searchCheck = React.useCallback(
		(searchIn?: string | null | false | 0) =>
			(searchIn || '').toLowerCase().includes(lowSearchText),
		[lowSearchText],
	)

	const persistentOptions = useSelector(selectPersistentOptions)
	const suggestions = Object.values(persistentOptions?.suggestions).filter(
		(i: any) => i.state === 'unread',
	)
	const configurations = Object.values(persistentOptions?.configurations).filter(
		(i: any) => i.state === 'unread',
	)
	const hasSuggestion: number = suggestions.length
	const hasConfigurations: number = configurations.length

	const conversationsDict = useConversationsDict()

	const searchConversations = React.useMemo(
		() =>
			searching
				? pickBy(
						conversationsDict,
						val =>
							val?.type === beapi.messenger.Conversation.Type.MultiMemberType &&
							searchCheck(val?.displayName),
				  )
				: {},
		[conversationsDict, searchCheck, searching],
	)

	const contacts = useContactsDict()

	const searchContacts = React.useMemo(
		() => (searching ? pickBy(contacts, (val: any) => searchCheck(val.displayName)) : {}),
		[contacts, searchCheck, searching],
	)

	const searchInteractions = React.useRef<beapi.messenger.IInteraction[]>([])
	const [earliestResult, setEarliestResult] = React.useState('')

	useEffect(() => {
		let canceled = false
		searchInteractions.current = []

		if (searchText.trim() === '') {
			return
		}

		;(async () => {
			await new Promise(resolve => {
				setTimeout(() => resolve(true), 200)
			})
			if (canceled) {
				return
			}

			try {
				let earliestResult = ''
				setEarliestResult('')
				while (true) {
					const results = await messengerClient?.messageSearch({
						query: searchText,
						refCid: earliestResult,
						limit: 10,
					})

					if (!results || results.results.length === 0) {
						canceled = true
					}

					if (canceled) {
						return
					}

					searchInteractions.current = searchInteractions.current.concat(results!.results)
					setEarliestResult(results!.results[results!.results.length - 1].cid!)
					earliestResult = results!.results[results!.results.length - 1].cid!
					// TODO: remove this loop, add loading on scroll
				}
			} catch (e) {
				canceled = true
				console.warn(e)
			}
		})()

		return () => {
			canceled = true
		}
	}, [messengerClient, searchInteractions, searchText])

	const hasResults = [searchConversations, searchContacts, searchInteractions.current].some(
		c => Object.keys(c).length > 0,
	)
	const styleBackground = useMemo(
		() =>
			requests.length > 0 && !searchText?.length
				? { backgroundColor: colors['background-header'] }
				: { backgroundColor: colors['main-background'] },
		[requests.length, searchText, colors],
	)

	return (
		<View style={[styleBackground, { flex: 1 }]}>
			<SafeAreaView>
				<StatusBar
					backgroundColor={
						requests.length && !isOnTop ? colors['background-header'] : colors['main-background']
					}
					barStyle={requests.length && !isOnTop ? 'light-content' : 'dark-content'}
				/>
			</SafeAreaView>
			<ScrollView
				ref={scrollRef}
				stickyHeaderIndices={!searchText?.length && !hasResults ? [1] : [0]}
				showsVerticalScrollIndicator={false}
				scrollEventThrottle={16}
				contentContainerStyle={{ flexGrow: 1 }}
				keyboardShouldPersistTaps={'handled'}
				onScrollEndDrag={e => {
					if (e.nativeEvent.contentOffset.y < 0) {
						setRefresh(true)
					}
				}}
				onScroll={e => {
					if (e.nativeEvent.contentOffset) {
						if (e.nativeEvent.contentOffset.y >= layoutRequests.height) {
							setIsOnTop(true)
						} else {
							setIsOnTop(false)
						}
					}
				}}
			>
				{!searchText?.length ? (
					<IncomingRequests items={requests} onLayout={onLayoutRequests} />
				) : null}
				<HomeHeader
					isOnTop={isOnTop}
					hasRequests={requests.length > 0}
					scrollRef={scrollRef}
					value={searchText}
					onChange={setSearchText}
					refresh={refresh}
					setRefresh={setRefresh}
					onLongPress={setIsLongPress}
					isMultiAccount={isLongPress}
				/>
				{searchText?.length ? (
					<>
						{(searchText.startsWith('https://berty.tech/id') ||
							searchText.startsWith('berty://')) && (
							<View style={[{ flexDirection: 'row', justifyContent: 'center' }]}>
								<View style={[border.shadow.large, border.radius.medium]}>
									<ButtonSettingV2
										text='Open Berty Link'
										icon='external-link-outline'
										onPress={() =>
											navigate('Chat.ManageDeepLink', { type: 'link', value: searchText })
										}
									/>
								</View>
							</View>
						)}
						<SearchComponent
							insets={null}
							conversations={searchConversations}
							contacts={searchContacts}
							interactions={searchInteractions.current}
							value={searchText}
							hasResults={hasResults}
							earliestInteractionCID={earliestResult}
						/>
					</>
				) : (
					<View style={{ height: '100%' }}>
						<Conversations
							items={conversations}
							suggestions={suggestions}
							configurations={configurations}
							addBot={setIsAddBot}
						/>
						{!hasConversations && !hasSuggestion && !hasConfigurations && (
							<View style={{ backgroundColor: colors['main-background'] }}>
								<View style={[flex.justify.center, flex.align.center, margin.top.scale(60)]}>
									<View>
										<EmptyChat width={350 * scaleSize} height={350 * scaleHeight} />
										<UnifiedText
											style={[
												text.align.center,
												text.color.grey,
												text.light,
												opacity(0.3),
												margin.top.big,
											]}
										>
											{t('main.home.no-contacts')}
										</UnifiedText>
									</View>
								</View>
							</View>
						)}
					</View>
				)}
			</ScrollView>
			<PrimaryFloatingButton onPress={() => navigate('Chat.Share')} />
			{isLongPress ? (
				<MultiAccount
					onPress={() => {
						setIsLongPress(false)
					}}
				/>
			) : null}
			{isAddBot.isVisible ? (
				<AddBot
					link={isAddBot.link}
					displayName={isAddBot.displayName}
					closeModal={() => setIsAddBot({ ...isAddBot, isVisible: false })}
				/>
			) : null}
		</View>
	)
}
