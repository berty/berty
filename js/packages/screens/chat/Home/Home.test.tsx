import { renderScreen } from '@berty/utils/testing/renderScreen.test'

import { Home } from './Home'

test('Chat.Home renders correctly', () => {
	const { toJSON } = renderScreen('Chat.Home', Home)
	expect(toJSON()).toMatchSnapshot()
})
