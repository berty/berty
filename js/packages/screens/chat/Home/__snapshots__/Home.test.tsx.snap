// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chat.Home renders correctly 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    Array [
      Object {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <View
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "backgroundColor": "#FFFFFF",
          },
          Object {
            "flex": 1,
          },
        ]
      }
    >
      <RCTSafeAreaView
        emulateUnlessSupported={true}
      />
      <RCTScrollView
        contentContainerStyle={
          Object {
            "flexGrow": 1,
          }
        }
        keyboardShouldPersistTaps="handled"
        onScroll={[Function]}
        onScrollEndDrag={[Function]}
        scrollEventThrottle={16}
        showsVerticalScrollIndicator={false}
        stickyHeaderIndices={
          Array [
            1,
          ]
        }
      >
        <View>
          <View>
            <View>
              <View
                style={
                  Array [
                    Object {
                      "borderTopLeftRadius": 32,
                      "borderTopRightRadius": 32,
                    },
                    Object {
                      "paddingHorizontal": 27,
                    },
                    Object {
                      "alignItems": "center",
                      "backgroundColor": "#FFFFFF",
                    },
                  ]
                }
              >
                <View
                  style={
                    Object {
                      "alignItems": "center",
                      "flexDirection": "row",
                      "paddingVertical": 15,
                    }
                  }
                >
                  <View
                    style={
                      Object {
                        "alignItems": "flex-end",
                        "flex": 1,
                        "marginLeft": 5,
                      }
                    }
                  >
                    <View
                      accessible={true}
                      focusable={true}
                      onClick={[Function]}
                      onResponderGrant={[Function]}
                      onResponderMove={[Function]}
                      onResponderRelease={[Function]}
                      onResponderTerminate={[Function]}
                      onResponderTerminationRequest={[Function]}
                      onStartShouldSetResponder={[Function]}
                      style={
                        Object {
                          "opacity": 1,
                        }
                      }
                    >
                      <LottieViewMock
                        loop={false}
                        sourceHash="+mcArBfINd658SwKHRJDKgdNzAVJXyDYK6qYNpx4RA0="
                        style={
                          Object {
                            "width": 40,
                          }
                        }
                      />
                    </View>
                  </View>
                  <View
                    style={
                      Array [
                        Object {
                          "flex": 12,
                        },
                        Object {
                          "marginLeft": 9,
                        },
                        Object {
                          "marginRight": 25,
                        },
                      ]
                    }
                  >
                    <View
                      accessible={true}
                      focusable={true}
                      onClick={[Function]}
                      onResponderGrant={[Function]}
                      onResponderMove={[Function]}
                      onResponderRelease={[Function]}
                      onResponderTerminate={[Function]}
                      onResponderTerminationRequest={[Function]}
                      onStartShouldSetResponder={[Function]}
                      style={
                        Object {
                          "alignItems": "center",
                          "backgroundColor": "#F2F2F2",
                          "borderRadius": 14,
                          "flexDirection": "row",
                          "height": 42,
                          "justifyContent": "flex-start",
                          "opacity": 1,
                          "paddingHorizontal": 12,
                        }
                      }
                    >
                      <View
                        style={
                          Array [
                            Object {
                              "alignItems": "stretch",
                              "flexDirection": "row",
                              "justifyContent": "space-evenly",
                            },
                          ]
                        }
                      >
                        <View
                          style={
                            Object {
                              "transform": Array [
                                Object {
                                  "scale": 1,
                                },
                              ],
                            }
                          }
                        >
                          <RNSVGMock
                            fill="#D0D0D6"
                            height={20}
                            viewBox="0 0 24 24"
                            width={20}
                          />
                        </View>
                      </View>
                      <View
                        style={
                          Array [
                            Object {
                              "marginLeft": 9,
                            },
                            Object {
                              "alignItems": "flex-start",
                              "flex": 6,
                              "flexDirection": "row",
                            },
                          ]
                        }
                      >
                        <TextInput
                          accessibilityLabel="main.home.input-placeholder"
                          allowFontScaling={true}
                          autoCapitalize="none"
                          autoCorrect={false}
                          iconName="search-outline"
                          onChangeText={[Function]}
                          placeholder="main.home.input-placeholder"
                          placeholderTextColor="#D0D0D6"
                          rejectResponderTermination={true}
                          style={
                            Array [
                              Object {
                                "fontSize": 30,
                              },
                              Object {
                                "color": "#393C63",
                                "flex": 1,
                                "fontFamily": "Open Sans",
                              },
                            ]
                          }
                          underlineColorAndroid="transparent"
                          value=""
                        />
                      </View>
                    </View>
                  </View>
                  <View
                    accessible={true}
                    focusable={true}
                    onClick={[Function]}
                    onResponderGrant={[Function]}
                    onResponderMove={[Function]}
                    onResponderRelease={[Function]}
                    onResponderTerminate={[Function]}
                    onResponderTerminationRequest={[Function]}
                    onStartShouldSetResponder={[Function]}
                    style={
                      Object {
                        "alignItems": "center",
                        "flex": 1,
                        "flexDirection": "row",
                        "justifyContent": "center",
                        "opacity": 1,
                      }
                    }
                  >
                    <View
                      style={
                        Object {
                          "zIndex": -1,
                        }
                      }
                    >
                      <View
                        style={
                          Array [
                            undefined,
                            Object {
                              "alignItems": "center",
                              "backgroundColor": "#FFFFFF",
                              "borderRadius": 17.5,
                              "height": 35,
                              "justifyContent": "center",
                              "width": 35,
                            },
                          ]
                        }
                      >
                        <View
                          style={
                            Object {
                              "alignItems": "center",
                              "justifyContent": "center",
                            }
                          }
                        >
                          <View
                            style={
                              Array [
                                undefined,
                                Object {
                                  "alignItems": "center",
                                  "backgroundColor": "#b8bd4a",
                                  "borderRadius": 17.5,
                                  "height": 35,
                                  "justifyContent": "center",
                                  "width": 35,
                                },
                              ]
                            }
                          >
                            <Text
                              style={
                                Array [
                                  Object {
                                    "fontSize": 30,
                                  },
                                  Object {
                                    "color": "#383B62",
                                    "fontFamily": "Open Sans",
                                  },
                                  Object {
                                    "color": "#FFFFFF",
                                    "fontSize": 17.5,
                                    "includeFontPadding": false,
                                  },
                                ]
                              }
                            >
                              ?
                            </Text>
                          </View>
                        </View>
                      </View>
                    </View>
                  </View>
                </View>
              </View>
            </View>
          </View>
          <View
            style={
              Object {
                "height": "100%",
              }
            }
          >
            <View
              style={
                Object {
                  "backgroundColor": "#FFFFFF",
                }
              }
            >
              <View
                style={
                  Array [
                    Object {
                      "justifyContent": "center",
                    },
                    Object {
                      "alignItems": "center",
                    },
                    Object {
                      "marginTop": 60,
                    },
                  ]
                }
              >
                <View>
                  <AssetMock
                    height={350}
                    width={350}
                  />
                  <Text
                    style={
                      Array [
                        Object {
                          "fontSize": 30,
                        },
                        Object {
                          "color": "#383B62",
                          "fontFamily": "Open Sans",
                        },
                        Array [
                          Object {
                            "textAlign": "center",
                          },
                          Object {
                            "color": "#979797",
                          },
                          Object {
                            "fontFamily": "Light Open Sans",
                          },
                          Object {
                            "opacity": 0.3,
                          },
                          Object {
                            "marginTop": 32,
                          },
                        ],
                      ]
                    }
                  >
                    main.home.no-contacts
                  </Text>
                </View>
              </View>
            </View>
          </View>
        </View>
      </RCTScrollView>
      <View
        style={
          Object {
            "bottom": 20,
            "position": "absolute",
            "right": 20,
          }
        }
      >
        <View
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "#535AE4",
              "borderRadius": 52,
              "elevation": 5,
              "height": 52,
              "justifyContent": "center",
              "marginBottom": 20,
              "opacity": 1,
              "shadowColor": "#000000",
              "shadowOffset": Object {
                "height": 5,
                "width": 0,
              },
              "shadowOpacity": 0.2,
              "shadowRadius": 10,
              "width": 52,
            }
          }
        >
          <View
            style={
              Object {
                "transform": Array [
                  Object {
                    "scale": 1,
                  },
                ],
              }
            }
          >
            <RNSVGMock
              fill="white"
              height={30}
              viewBox="0 0 24 24"
              width={30}
            />
          </View>
        </View>
      </View>
    </View>
  </View>
</RNCSafeAreaProvider>
`;
